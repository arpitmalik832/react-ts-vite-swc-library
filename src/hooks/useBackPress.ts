import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import {
  clearStack,
  popStack,
  pushStack,
} from '../redux/slices/navigationSlice';
import beforeUnload from '../utils/eventListeners/beforeUnload';
import { log } from '../utils/logsUtils';
import { ReduxState, NavigationRedux, VoidFunction } from '../types/types.d';

function useBackPress() {
  const { stack } = useSelector<ReduxState, NavigationRedux>(
    state => state.navigation,
  );
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleBackPress = useCallback(() => {
    if (stack.length) {
      dispatch(popStack(undefined));
    } else {
      navigate(-1);
    }
  }, [stack]);

  window.backPress = handleBackPress;

  useEffect(() => {
    beforeUnload.subscribe(() => {
      log('ðŸ˜¬ user back clicked!!');
    });

    return () => {
      beforeUnload.unSubscribe();
    };
  }, []);

  function push(callback: VoidFunction) {
    dispatch(pushStack(callback));
  }

  function pop() {
    handleBackPress();
  }

  const clear = useCallback(() => {
    if (stack.length) {
      dispatch(clearStack(undefined));
    }
  }, [stack]);

  return { stack, push, pop, clear };
}

export default useBackPress;
