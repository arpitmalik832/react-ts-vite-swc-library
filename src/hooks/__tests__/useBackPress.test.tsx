import { useEffect } from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import {
  configureStore,
  createSlice,
  SliceCaseReducers,
  SliceSelectors,
} from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import useBackPress from '../useBackPress';
import {
  BeforeUnloadEventListener,
  NavigationRedux,
} from '../../types/types.d';

jest.mock('react-router-dom', () => ({
  __esModule: true,
  useNavigate: jest.fn().mockReturnValue(jest.fn()),
}));

jest.mock('../../utils/eventListeners/beforeUnload', () => ({
  __esModule: true,
  default: {
    subscribe: (e: BeforeUnloadEventListener) => {
      e.call(window, {} as BeforeUnloadEvent);
    },
    unSubscribe: jest.fn(),
  },
}));

jest.mock('../../utils/commonUtils', () => ({
  __esModule: true,
  log: jest.fn(),
}));

describe('useBackPress unit tests', () => {
  it('snapshot test', () => {
    const navigationSlice = createSlice<
      NavigationRedux,
      SliceCaseReducers<NavigationRedux>,
      string,
      SliceSelectors<NavigationRedux>,
      string
    >({
      name: 'navigation',
      initialState: { stack: [] },
      reducers: {},
    });

    const store = configureStore({
      reducer: {
        navigation: navigationSlice.reducer,
      },
    });

    function TempComponent() {
      useBackPress();

      return <div data-testid="temp-component" />;
    }

    const component = render(
      <Provider store={store}>
        <TempComponent />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });

  it('testing functions', () => {
    const navigationSlice = createSlice<
      NavigationRedux,
      SliceCaseReducers<NavigationRedux>,
      string,
      SliceSelectors<NavigationRedux>,
      string
    >({
      name: 'navigation',
      initialState: {
        stack: [],
      },
      reducers: {
        pushStack: (state, action) => {
          state.stack.push(action.payload);
          return state;
        },
        popStack: state => {
          const top = state.stack.pop();
          if (top) {
            top();
          }
          return state;
        },
        clearStack: state => ({
          ...state,
          stack: [],
        }),
      },
    });

    const store = configureStore({
      reducer: {
        navigation: navigationSlice.reducer,
      },
    });

    function TempComponent() {
      const { push, pop } = useBackPress();

      useEffect(() => {
        push(jest.fn());
      }, []);

      return (
        <div data-testid="temp-component">
          <button type="button" data-testid="temp-btn" onClick={pop}>
            temp btn
          </button>
        </div>
      );
    }

    const { getByTestId } = render(
      <Provider store={store}>
        <TempComponent />
      </Provider>,
    );

    fireEvent.click(getByTestId('temp-btn'));
    fireEvent.click(getByTestId('temp-btn'));
  });

  it('testing functions', () => {
    const navigationSlice = createSlice<
      NavigationRedux,
      SliceCaseReducers<NavigationRedux>,
      string,
      SliceSelectors<NavigationRedux>,
      string
    >({
      name: 'navigation',
      initialState: {
        stack: [],
      },
      reducers: {
        pushStack: (state, action) => {
          state.stack.push(action.payload);
          return state;
        },
        popStack: state => {
          const top = state.stack.pop();
          if (top) {
            top();
          }
          return state;
        },
        clearStack: state => ({
          ...state,
          stack: [],
        }),
      },
    });

    const store = configureStore({
      reducer: {
        navigation: navigationSlice.reducer,
      },
    });

    function TempComponent() {
      const { push, clear } = useBackPress();

      useEffect(() => {
        push(jest.fn());
      }, []);

      return (
        <div data-testid="temp-component">
          <button type="button" data-testid="temp-btn" onClick={clear}>
            temp btn
          </button>
        </div>
      );
    }

    const { getByTestId } = render(
      <Provider store={store}>
        <TempComponent />
      </Provider>,
    );

    fireEvent.click(getByTestId('temp-btn'));
    fireEvent.click(getByTestId('temp-btn'));
  });
});
