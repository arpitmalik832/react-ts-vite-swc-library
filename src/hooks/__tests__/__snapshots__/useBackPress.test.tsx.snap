// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBackPress unit tests snapshot test 1`] = `
{
  "clear": [Function],
  "pop": [Function],
  "push": [Function],
  "stack": [],
}
`;
